src/main/scala/chisel3/simulator/PeekPokeAPI.scala

```diff
@@ -127,58 +304,236 @@ trait PeekPokeAPI {
+
+    def expect(expected: T, allowPartial: Boolean)(implicit sourceInfo: SourceInfo): Unit =
+      expect(expected, (obs: T, exp: T) => "", allowPartial)
+
+    override def expect(expected: T)(implicit sourceInfo: SourceInfo): Unit =
+      expect(expected, (obs: T, exp: T) => "", allowPartial = false)
+  }
+
+  implicit class TestableVec[T <: Data](val data: Vec[T]) extends PeekPokable[Vec[T]] {
+    override def peek(): Vec[T] = {
+      val elementValues = data.getElements.map(_.peek().asInstanceOf[T])
+      chiselTypeOf(data).Lit(elementValues.zipWithIndex.map { _.swap }: _*)
+    }
+
+    override def poke(literal: Vec[T]): Unit = {
+      require(data.length == literal.length, s"Vec length mismatch: expected ${data.length}, got ${literal.length}")
+      data.getElements.zip(literal.getElements).foreach {
+        case (portEl, valueEl) if portEl.getClass == valueEl.getClass =>
+          portEl.poke(valueEl)
+        case (portEl, valueEl) =>
+          throw new Exception(
+            s"This should never happen! Port element type: ${portEl.getClass} != literal element ${valueEl.getClass}"
           )
+      }
+    }
+
+    final def expectVec[U <: Data](expected: Vec[U], buildMessage: (Vec[T], Vec[U], Int) => String)(
+      implicit sourceInfo: SourceInfo
+    ): Unit = {
+      require(
+        expected.length == data.length,
+        s"Vec length mismatch: Data port has ${data.length} elements while the expected value is of length ${expected.length}"
+      )
+      val peekedValue = data.peek()
+      data.zip(expected).zipWithIndex.foreach { case ((datEl, expEl), index) =>
+        expEl match {
+          case DontCare =>
+          // TODO: missing elements?
+          case exp if exp.getClass == datEl.getClass =>
+            datEl.expect(
+              exp.asInstanceOf[T],
+              (obs: T, exp: T) => buildMessage(peekedValue, expected, index)
+            )
+          case _ =>
+            throw new Exception(
+              s"Type mismatch: ${expEl.getClass} != ${datEl.getClass}"
+            )
         }
       }
     }
+
+    override def expect(expected: Vec[T], buildMessage: (Vec[T], Vec[T]) => String)(
+      implicit sourceInfo: SourceInfo
+    ): Unit =
+      expectVec[T](
+        expected,
+        (observed: Vec[T], expected: Vec[T], idx: Int) =>
+          buildMessage(observed, expected) + s". First mismatch at index $idx."
+      )
+  }
+
+  implicit class TestableData[T <: Data](val data: T) extends PeekPokable[T] {
+
+    def peek(): T = {
+      data match {
+        case x: Bool     => new TestableBool(x).peek().asInstanceOf[T]
+        case x: UInt     => new TestableUInt(x).peek().asInstanceOf[T]
+        case x: SInt     => new TestableSInt(x).peek().asInstanceOf[T]
+        case x: EnumType => new TestableEnum(x).peek().asInstanceOf[T]
+        case x: Record   => new TestableRecord(x).peek().asInstanceOf[T]
+        case x: Vec[_]   => new TestableVec(x).peek().asInstanceOf[T]
```

Comment:
Are these the `asInstanceOfs` you're concerned about? I actually think there are fine--we know the type is correct because we have matched on the concrete runtime type, so barring a bug in `TestableBool` (or whatever for the given type), the cast will always work. The issue is just convincing the compiler that the type is correct--which it is.

Reply:
Yes, my primary concern was really `Vec`, but I agree that all should be fine.